File: backend/scripts/auto_webhook_manager.py
#!/usr/bin/env python3

"""
Auto Webhook Manager (Python Version)

This script:
1. Starts ngrok tunnel
2. Monitors for URL changes
3. Updates .env files
4. Updates Twilio webhook
5. Updates Exotel webhook

Usage:
    python auto_webhook_manager.py
    
Or with specific port:
    python auto_webhook_manager.py --port 3000

Requirements:
    pip install requests python-dotenv

Environment variables (can be in .env file):
    - TWILIO_ACCOUNT_SID
    - TWILIO_AUTH_TOKEN
    - TWILIO_PHONE_NUMBER
    - EXOTEL_SID
    - EXOTEL_TOKEN
    - EXOTEL_APP_ID
    - EXOTEL_WEBHOOK_NUMBER
    - SLACK_WEBHOOK_URL (optional)
"""

import os
import sys
import time
import json
import signal
import subprocess
import argparse
import requests
from datetime import datetime
from pathlib import Path
from urllib.parse import urlparse

try:
    from dotenv import load_dotenv
    load_dotenv()
except ImportError:
    pass

try:
    import requests
except ImportError:
    print("Please install requests: pip install requests")
    sys.exit(1)

CONFIG = {
    'ngrok_port': 3000,
    'ngrok_region': 'in',
    'check_interval': 15,
    'ngrok_api_url': 'http://127.0.0.1:4040/api/tunnels'
}

COLORS = {
    'reset': '\033[0m',
    'green': '\033[32m',
    'red': '\033[31m',
    'yellow': '\033[33m',
    'blue': '\033[34m',
    'cyan': '\033[36m'
}

class WebhookManager:
    def __init__(self, port=3000, region='in'):
        self.ngrok_port = port
        self.ngrok_region = region
        self.current_url = None
        self.ngrok_process = None
        self.running = False
        
    def log(self, msg, color='reset'):
        timestamp = datetime.now().strftime('%H:%M:%S')
        print(f"{COLORS['cyan']}[{timestamp}]{COLORS['reset']} {COLORS[color]}{msg}{COLORS['reset']}")
        
    def check_ngrok_running(self):
        """Check if ngrok is already running"""
        try:
            response = requests.get(CONFIG['ngrok_api_url'], timeout=2)
            return response.json()
        except:
            return None
            
    def start_ngrok(self):
        """Start ngrok tunnel"""
        self.log("Starting ngrok tunnel...", 'blue')
        
        try:
            self.ngrok_process = subprocess.Popen(
                ['ngrok', 'http', '--region', self.ngrok_region, '--addr', str(self.ngrok_port)],
                stdout=subprocess.PIPE,
                stderr=subprocess.PIPE
            )
            
            for _ in range(20):
                time.sleep(1.5)
                tunnels = self.check_ngrok_running()
                if tunnels and tunnels.get('tunnels'):
                    self.log("Ngrok started successfully", 'green')
                    return tunnels
                    
            raise Exception("Ngrok failed to start within timeout")
        except FileNotFoundError:
            self.log("ngrok not found. Please install ngrok from https://ngrok.com/download", 'red')
            sys.exit(1)
            
    def get_ngrok_url(self):
        """Get current ngrok HTTPS URL"""
        tunnels = self.check_ngrok_running()
        if not tunnels:
            raise Exception("Ngrok not running")
            
        for tunnel in tunnels.get('tunnels', []):
            if tunnel.get('proto') == 'https':
                return tunnel.get('public_url', '').rstrip('/')
                
        raise Exception("No HTTPS tunnel found")
        
    def update_env_files(self, base_url):
        """Update all .env files with new webhook URLs"""
        
        env_vars = {
            'WEBHOOK_BASE_URL': base_url,
            'VOICE_WEBHOOK': f"{base_url}/api/voice/incoming",
            'SMS_WEBHOOK': f"{base_url}/api/sms/incoming",
            'CALLBACK_URL': f"{base_url}/api/voice/callback",
            'STATUS_CALLBACK': f"{base_url}/api/voice/status"
        }
        
        env_paths = [
            Path(__file__).parent.parent / '.env',
            Path(__file__).parent.parent.parent / '.env',
        ]
        
        for env_path in env_paths:
            if not env_path.exists():
                continue
                
            content = env_path.read_text()
            
            for key, value in env_vars.items():
                lines = content.split('\n')
                found = False
                for i, line in enumerate(lines):
                    if line.strip().startswith(f'{key}='):
                        lines[i] = f'{key}={value}'
                        found = True
                        break
                if not found:
                    lines.append(f'{key}={value}')
                    
                content = '\n'.join(lines)
                env_path.write_text(content)
                self.log(f"Updated .env: {env_path.name}", 'green')
                
    def update_twilio_webhook(self, base_url):
        """Update Twilio phone number webhook"""
        
        twilio_sid = os.getenv('TWILIO_ACCOUNT_SID')
        twilio_auth = os.getenv('TWILIO_AUTH_TOKEN')
        twilio_number = os.getenv('TWILIO_PHONE_NUMBER')
        
        if not all([twilio_sid, twilio_auth, twilio_number]):
            self.log("Twilio credentials not found, skipping...", 'yellow')
            return {'success': False, 'reason': 'missing_credentials'}
            
        try:
            self.log("Updating Twilio webhook...", 'blue')
            
            auth = (twilio_sid, twilio_auth)
            voice_url = f"{base_url}/api/voice/incoming"
            
            # Get phone numbers
            resp = requests.get(
                f"https://api.twilio.com/2010-04-01/Accounts/{twilio_sid}/IncomingPhoneNumbers.json",
                auth=auth
            )
            numbers = resp.json().get('incoming_phone_numbers', [])
            
            # Find our number
            my_number = None
            for num in numbers:
                if twilio_number in num.get('phone_number', ''):
                    my_number = num
                    break
                    
            if not my_number:
                self.log(f"Twilio number {twilio_number} not found", 'red')
                return {'success': False, 'reason': 'number_not_found'}
                
            # Update webhook
            resp = requests.post(
                f"https://api.twilio.com/2010-04-01/Accounts/{twilio_sid}/IncomingPhoneNumbers/{my_number['sid']}.json",
                auth=auth,
                data={
                    'voice_url': voice_url,
                    'voice_method': 'POST',
                    'sms_url': f"{base_url}/api/sms/incoming",
                    'sms_method': 'POST'
                }
            )
            
            if resp.status_code in [200, 201]:
                self.log(f"✅ Twilio webhook updated: {voice_url}", 'green')
                return {'success': True, 'url': voice_url}
            else:
                self.log(f"Twilio error: {resp.text}", 'red')
                return {'success': False, 'error': resp.text}
                
        except Exception as e:
            self.log(f"Twilio error: {str(e)}", 'red')
            return {'success': False, 'error': str(e)}
            
    def update_exotel_webhook(self, base_url):
        """Update Exotel app webhook"""
        
        exotel_sid = os.getenv('EXOTEL_SID')
        exotel_token = os.getenv('EXOTEL_TOKEN')
        exotel_app_id = os.getenv('EXOTEL_APP_ID')
        
        if not all([exotel_sid, exotel_token, exotel_app_id]):
            self.log("Exotel credentials not found, skipping...", 'yellow')
            return {'success': False, 'reason': 'missing_credentials'}
            
        try:
            self.log("Updating Exotel webhook...", 'blue')
            
            voice_url = f"{base_url}/api/voice/incoming"
            
            resp = requests.put(
                f"https://api.exotel.com/v1/Accounts/{exotel_sid}/apps/{exotel_app_id}",
                auth=(exotel_sid, exotel_token),
                json={
                    'VoiceCallbackUrl': voice_url,
                    'HangupCallbackUrl': f"{base_url}/api/voice/hangup",
                    'StatusCallbackUrl': f"{base_url}/api/voice/status"
                }
            )
            
            if resp.status_code in [200, 201]:
                self.log(f"✅ Exotel webhook updated: {voice_url}", 'green')
                return {'success': True, 'url': voice_url}
            else:
                self.log(f"Exotel error: {resp.text}", 'red')
                return {'success': False, 'error': resp.text}
                
        except Exception as e:
            self.log(f"Exotel error: {str(e)}", 'red')
            return {'success': False, 'error': str(e)}
            
    def update_exotel_manual(self, base_url):
        """Update Exotel manual portal number webhook"""
        
        exotel_sid = os.getenv('EXOTEL_SID')
        exotel_token = os.getenv('EXOTEL_TOKEN')
        exotel_webhook_number = os.getenv('EXOTEL_WEBHOOK_NUMBER')
        
        if not all([exotel_sid, exotel_token, exotel_webhook_number]):
            self.log("Exotel Manual credentials not found, skipping...", 'yellow')
            return {'success': False, 'reason': 'missing_credentials'}
            
        try:
            self.log("Updating Exotel Manual Number webhook...", 'blue')
            
            voice_url = f"{base_url}/api/voice/incoming"
            
            resp = requests.put(
                f"https://api.exotel.com/v1/Accounts/{exotel_sid}/incomingphonenumbers/{exotel_webhook_number}",
                auth=(exotel_sid, exotel_token),
                json={
                    'VoiceCallbackUrl': voice_url,
                    'HangupCallbackUrl': f"{base_url}/api/voice/hangup"
                }
            )
            
            if resp.status_code in [200, 201]:
                self.log(f"✅ Exotel Manual webhook updated: {voice_url}", 'green')
                return {'success': True, 'url': voice_url}
            else:
                self.log(f"Exotel Manual error: {resp.text}", 'red')
                return {'success': False, 'error': resp.text}
                
        except Exception as e:
            self.log(f"Exotel Manual error: {str(e)}", 'red')
            return {'success': False, 'error': str(e)}
            
    def update_all(self, base_url):
        """Update all webhooks"""
        
        print()
        print("=" * 50)
        self.log(f"Updating all webhooks to: {base_url}", 'blue')
        print("=" * 50)
        
        self.update_env_files(base_url)
        
        results = {
            'twilio': self.update_twilio_webhook(base_url),
            'exotel': self.update_exotel_webhook(base_url),
            'exotel_manual': self.update_exotel_manual(base_url)
        }
        
        print()
        self.log("✅ Webhook update complete!", 'green')
        return results
        
    def monitor(self):
        """Monitor for URL changes"""
        
        if not self.running:
            return
            
        try:
            new_url = self.get_ngrok_url()
            
            if new_url != self.current_url:
                self.log(f"URL changed: {self.current_url} → {new_url}", 'yellow')
                self.current_url = new_url
                self.update_all(new_url)
                
        except Exception as e:
            self.log(f"Monitor error: {str(e)}", 'red')
            
    def run(self):
        """Main run loop"""
        
        print("""
╔════════════════════════════════════════════════════════════╗
║          🌐 Auto Webhook Manager (Python)                  ║
║          Updates ngrok webhooks automatically              ║
╚════════════════════════════════════════════════════════════╝
        """)
        
        try:
            tunnels = self.check_ngrok_running()
            
            if not tunnels or not tunnels.get('tunnels'):
                self.log("Ngrok not running, starting...", 'yellow')
                tunnels = self.start_ngrok()
                
            self.current_url = self.get_ngrok_url()
            self.log(f"Current URL: {self.current_url}", 'green')
            
            self.update_all(self.current_url)
            
            self.log(f"\nMonitoring for URL changes every {CONFIG['check_interval']}s...", 'blue')
            self.log("Press Ctrl+C to stop\n", 'yellow')
            
            self.running = True
            
            while self.running:
                time.sleep(CONFIG['check_interval'])
                self.monitor()
                
        except KeyboardInterrupt:
            self.log("\nShutting down...", 'yellow')
            self.running = False
            if self.ngrok_process:
                self.ngrok_process.terminate()
                
        except Exception as e:
            self.log(f"Fatal error: {str(e)}", 'red')
            sys.exit(1)


def main():
    parser = argparse.ArgumentParser(description='Auto Webhook Manager')
    parser.add_argument('--port', type=int, default=3000, help='Port for ngrok tunnel')
    parser.add_argument('--region', type=str, default='in', help='ngrok region (in, us, eu, etc.)')
    parser.add_argument('--interval', type=int, default=15, help='Check interval in seconds')
    
    args = parser.parse_args()
    
    CONFIG['ngrok_port'] = args.port
    CONFIG['ngrok_region'] = args.region
    CONFIG['check_interval'] = args.interval
    
    manager = WebhookManager(port=args.port, region=args.region)
    manager.run()


if __name__ == '__main__':
    main()

File: backend/scripts/auto-webhook-manager.js
#!/usr/bin/env node

/**
 * Auto Webhook Manager
 * 
 * This script:
 * 1. Starts ngrok tunnel
 * 2. Monitors for URL changes
 * 3. Updates .env files
 * 4. Updates Twilio webhook
 * 5. Updates Exotel webhook
 * 6. Sends notifications
 * 
 * Usage: node auto-webhook-manager.js
 * 
 * Environment variables required in .env:
 * - TWILIO_ACCOUNT_SID
 * - TWILIO_AUTH_TOKEN  
 * - TWILIO_PHONE_NUMBER
 * - EXOTEL_SID
 * - EXOTEL_TOKEN
 * - EXOTEL_APP_ID
 * - EXOTEL_WEBHOOK_NUMBER
 * - SLACK_WEBHOOK_URL (optional)
 */

const { spawn, exec } = require('child_process');
const fs = require('fs');
const path = require('path');
const axios = require('axios');
require('dotenv').config({ path: require('path').resolve(__dirname, '../.env') });

const CONFIG = {
    NGROK_PORT: process.env.NGROK_PORT || 3000,
    NGROK_REGION: process.env.NGROK_REGION || 'in',
    CHECK_INTERVAL: 15000,
    NGROK_API_URL: 'http://127.0.0.1:4040/api/tunnels'
};

let currentUrl = null;
let ngrokProcess = null;
let isRunning = false;

const COLORS = {
    reset: '\x1b[0m',
    green: '\x1b[32m',
    red: '\x1b[31m',
    yellow: '\x1b[33m',
    blue: '\x1b[34m',
    cyan: '\x1b[36m'
};

function log(msg, color = 'reset') {
    const timestamp = new Date().toISOString().split('T')[1].split('.')[0];
    console.log(`${COLORS.cyan}[${timestamp}]${COLORS.reset} ${COLORS[color]}${msg}${COLORS.reset}`);
}

function sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
}

async function checkNgrokRunning() {
    try {
        const response = await axios.get(CONFIG.NGROK_API_URL, { timeout: 2000 });
        return response.data;
    } catch (err) {
        return null;
    }
}

async function startNgrok() {
    return new Promise((resolve, reject) => {
        log('Starting ngrok tunnel...', 'blue');
        
        ngrokProcess = spawn('ngrok', [
            'http',
            '--region', CONFIG.NGROK_REGION,
            '--addr', CONFIG.NGROK_PORT,
            '--log', 'stdout'
        ], {
            stdio: ['ignore', 'pipe', 'pipe'],
            detached: false
        });

        let startupOutput = '';
        
        ngrokProcess.stdout.on('data', (data) => {
            const output = data.toString();
            startupOutput += output;
            
            if (output.includes('started tunnel') || output.includes('Session Status')) {
                log('Ngrok started successfully', 'green');
            }
        });

        ngrokProcess.stderr.on('data', (data) => {
            const output = data.toString();
            if (output.includes('started tunnel') || output.includes('url=')) {
                log('Ngrok tunnel established', 'green');
            }
        });

        ngrokProcess.on('error', (err) => {
            log(`Ngrok error: ${err.message}`, 'red');
            reject(err);
        });

        let attempts = 0;
        const checkInterval = setInterval(async () => {
            attempts++;
            const tunnels = await checkNgrokRunning();
            if (tunnels && tunnels.tunnels && tunnels.tunnels.length > 0) {
                clearInterval(checkInterval);
                resolve(tunnels);
            }
            if (attempts > 20) {
                clearInterval(checkInterval);
                reject(new Error('Ngrok failed to start within timeout'));
            }
        }, 1500);
    });
}

async function getNgrokUrl() {
    const tunnels = await checkNgrokRunning();
    if (!tunnels) {
        throw new Error('Ngrok not running');
    }
    
    const httpsTunnel = tunnels.tunnels.find(t => t.proto === 'https');
    if (!httpsTunnel) {
        throw new Error('No HTTPS tunnel found');
    }
    
    return httpsTunnel.public_url.replace(/\/+$/, '');
}

function updateEnvFiles(baseUrl) {
    const envVars = {
        WEBHOOK_BASE_URL: baseUrl,
        VOICE_WEBHOOK: `${baseUrl}/api/voice/incoming`,
        SMS_WEBHOOK: `${baseUrl}/api/sms/incoming`,
        CALLBACK_URL: `${baseUrl}/api/voice/callback`,
        STATUS_CALLBACK: `${baseUrl}/api/voice/status`
    };

    const envPaths = [
        path.resolve(__dirname, '../.env'),
        path.resolve(__dirname, '../../.env'),
        path.resolve(__dirname, '../../../.env')
    ];

    envPaths.forEach(envPath => {
        if (!fs.existsSync(envPath)) return;
        
        let content = fs.readFileSync(envPath, 'utf8');
        
        Object.entries(envVars).forEach(([key, value]) => {
            const regex = new RegExp(`^${key}=.*$`, 'm');
            if (regex.test(content)) {
                content = content.replace(regex, `${key}=${value}`);
            } else {
                content += `\n${key}=${value}`;
            }
        });
        
        fs.writeFileSync(envPath, content);
        log(`Updated .env: ${path.basename(path.dirname(envPath))}/.env`, 'green');
    });
}

async function updateTwilioWebhook(baseUrl) {
    const TWILIO_SID = process.env.TWILIO_ACCOUNT_SID;
    const TWILIO_AUTH = process.env.TWILIO_AUTH_TOKEN;
    const TWILIO_NUMBER = process.env.TWILIO_PHONE_NUMBER;

    if (!TWILIO_SID || !TWILIO_AUTH || !TWILIO_NUMBER) {
        log('Twilio credentials not found, skipping...', 'yellow');
        return { success: false, reason: 'missing_credentials' };
    }

    try {
        log('Updating Twilio webhook...', 'blue');
        
        const auth = Buffer.from(`${TWILIO_SID}:${TWILIO_AUTH}`).toString('base64');
        
        const numbers = await axios.get(
            `https://api.twilio.com/2010-04-01/Accounts/${TWILIO_SID}/IncomingPhoneNumbers.json`,
            { headers: { Authorization: `Basic ${auth}` } }
        );

        const myNumber = numbers.data.incoming_phone_numbers.find(
            n => n.phone_number === TWILIO_NUMBER || 
                 n.phone_number.includes(TWILIO_NUMBER.replace('+', ''))
        );

        if (!myNumber) {
            log(`Twilio number ${TWILIO_NUMBER} not found`, 'red');
            return { success: false, reason: 'number_not_found' };
        }

        const voiceUrl = `${baseUrl}/api/voice/incoming`;
        
        await axios.post(
            `https://api.twilio.com/2010-04-01/Accounts/${TWILIO_SID}/IncomingPhoneNumbers/${myNumber.sid}.json`,
            new URLSearchParams({
                voice_url: voiceUrl,
                voice_method: 'POST',
                sms_url: `${baseUrl}/api/sms/incoming`,
                sms_method: 'POST'
            }),
            { 
                headers: { 
                    Authorization: `Basic ${auth}`, 
                    'Content-Type': 'application/x-www-form-urlencoded' 
                } 
            }
        );

        log(`✅ Twilio webhook updated: ${voiceUrl}`, 'green');
        return { success: true, url: voiceUrl };
        
    } catch (err) {
        log(`Twilio error: ${err.response?.data?.message || err.message}`, 'red');
        return { success: false, error: err.message };
    }
}

async function updateExotelWebhook(baseUrl) {
    const EXOTEL_SID = process.env.EXOTEL_SID;
    const EXOTEL_TOKEN = process.env.EXOTEL_TOKEN;
    const EXOTEL_APP_ID = process.env.EXOTEL_APP_ID;

    if (!EXOTEL_SID || !EXOTEL_TOKEN || !EXOTEL_APP_ID) {
        log('Exotel credentials not found, skipping...', 'yellow');
        return { success: false, reason: 'missing_credentials' };
    }

    try {
        log('Updating Exotel webhook...', 'blue');
        
        const voiceUrl = `${baseUrl}/api/voice/incoming`;
        
        const response = await axios.put(
            `https://api.exotel.com/v1/Accounts/${EXOTEL_SID}/apps/${EXOTEL_APP_ID}`,
            {
                VoiceCallbackUrl: voiceUrl,
                HangupCallbackUrl: `${baseUrl}/api/voice/hangup`,
                StatusCallbackUrl: `${baseUrl}/api/voice/status`
            },
            {
                auth: {
                    username: EXOTEL_SID,
                    password: EXOTEL_TOKEN
                },
                headers: {
                    'Content-Type': 'application/json'
                }
            }
        );

        log(`✅ Exotel webhook updated: ${voiceUrl}`, 'green');
        return { success: true, url: voiceUrl };
        
    } catch (err) {
        log(`Exotel error: ${err.response?.data?.message || err.message}`, 'red');
        return { success: false, error: err.message };
    }
}

async function updateExotelManualPortal(baseUrl) {
    const EXOTEL_SID = process.env.EXOTEL_SID;
    const EXOTEL_TOKEN = process.env.EXOTEL_TOKEN;
    const EXOTEL_WEBHOOK_NUMBER = process.env.EXOTEL_WEBHOOK_NUMBER;

    if (!EXOTEL_SID || !EXOTEL_TOKEN || !EXOTEL_WEBHOOK_NUMBER) {
        log('Exotel manual portal credentials not found, skipping...', 'yellow');
        return { success: false, reason: 'missing_credentials' };
    }

    try {
        log('Updating Exotel Manual Number webhook...', 'blue');
        
        const voiceUrl = `${baseUrl}/api/voice/incoming`;
        
        const response = await axios.put(
            `https://api.exotel.com/v1/Accounts/${EXOTEL_SID}/incomingphonenumbers/${EXOTEL_WEBHOOK_NUMBER}`,
            {
                VoiceCallbackUrl: voiceUrl,
                HangupCallbackUrl: `${baseUrl}/api/voice/hangup`
            },
            {
                auth: {
                    username: EXOTEL_SID,
                    password: EXOTEL_TOKEN
                },
                headers: {
                    'Content-Type': 'application/json'
                }
            }
        );

        log(`✅ Exotel Manual Number webhook updated: ${voiceUrl}`, 'green');
        return { success: true, url: voiceUrl };
        
    } catch (err) {
        log(`Exotel Manual Number error: ${err.response?.data?.message || err.message}`, 'red');
        return { success: false, error: err.message };
    }
}

async function sendSlackNotification(baseUrl, results) {
    const SLACK_WEBHOOK = process.env.SLACK_WEBHOOK_URL;
    
    if (!SLACK_WEBHOOK) return;
    
    try {
        const message = {
            text: '🔗 Ngrok Webhook Updated',
            blocks: [
                {
                    type: 'header',
                    text: { type: 'plain_text', text: '🔗 Webhook URL Updated' }
                },
                {
                    type: 'section',
                    fields: [
                        { type: 'mrkdwn', text: `*New URL:*\n${baseUrl}` },
                        { type: 'mrkdwn', text: `*Time:*\n${new Date().toLocaleString()}` }
                    ]
                },
                {
                    type: 'section',
                    text: { 
                        type: 'mrkdwn', 
                        text: `*Results:*\n• Twilio: ${results.twilio?.success ? '✅' : '❌'}\n• Exotel: ${results.exotel?.success ? '✅' : '❌'}\n• Env: ✅` 
                    }
                }
            ]
        };
        
        await axios.post(SLACK_WEBHOOK, message);
        log('Slack notification sent', 'green');
    } catch (err) {
        log(`Slack notification failed: ${err.message}`, 'yellow');
    }
}

async function updateAllWebhooks(baseUrl) {
    log(`\n${'='.repeat(50)}`, 'cyan');
    log(`Updating all webhooks to: ${baseUrl}`, 'blue');
    log('='.repeat(50), 'cyan');
    
    updateEnvFiles(baseUrl);
    
    const results = {
        twilio: await updateTwilioWebhook(baseUrl),
        exotel: await updateExotelWebhook(baseUrl),
        exotelManual: await updateExotelManualPortal(baseUrl)
    };
    
    await sendSlackNotification(baseUrl, results);
    
    log('\n✅ Webhook update complete!', 'green');
    return results;
}

async function monitorNgrok() {
    if (!isRunning) return;
    
    try {
        const newUrl = await getNgrokUrl();
        
        if (newUrl !== currentUrl) {
            log(`URL changed: ${currentUrl} → ${newUrl}`, 'yellow');
            currentUrl = newUrl;
            await updateAllWebhooks(newUrl);
        }
    } catch (err) {
        log(`Monitor error: ${err.message}`, 'red');
    }
}

async function main() {
    console.log(`
╔════════════════════════════════════════════════════════════╗
║          🌐 Auto Webhook Manager                           ║
║          Updates ngrok webhooks automatically              ║
╚════════════════════════════════════════════════════════════╝
    `);

    try {
        let tunnels = await checkNgrokRunning();
        
        if (!tunnels || !tunnels.tunnels || tunnels.tunnels.length === 0) {
            log('Ngrok not running, starting...', 'yellow');
            tunnels = await startNgrok();
        }
        
        currentUrl = await getNgrokUrl();
        log(`Current URL: ${currentUrl}`, 'green');
        
        await updateAllWebhooks(currentUrl);
        
        log(`\nMonitoring for URL changes every ${CONFIG.CHECK_INTERVAL/1000}s...`, 'blue');
        log('Press Ctrl+C to stop\n', 'yellow');
        
        isRunning = true;
        setInterval(monitorNgrok, CONFIG.CHECK_INTERVAL);
        
    } catch (err) {
        log(`Fatal error: ${err.message}`, 'red');
        process.exit(1);
    }
}

process.on('SIGINT', () => {
    log('\nShutting down...', 'yellow');
    isRunning = false;
    if (ngrokProcess) {
        ngrokProcess.kill();
    }
    process.exit(0);
});

process.on('SIGTERM', () => {
    log('\nShutting down...', 'yellow');
    isRunning = false;
    if (ngrokProcess) {
        ngrokProcess.kill();
    }
    process.exit(0);
});

main();
